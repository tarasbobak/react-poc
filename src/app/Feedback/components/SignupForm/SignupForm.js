import React, { Component } from 'react';
import cx from 'classnames';
import { Formik, Form, Field } from 'formik';
import PropTypes from 'prop-types';
import Yup from 'yup';
import InputFeedback from './InputFeedback/InputFeedback';
import styles from './SignupForm.scss';

const CN = styles['signup-form'];

const VALIDATION_SCHEMA = Yup.object().shape({
  email: Yup.string()
    .required('Please provide your email address, formatted as you@domain.com.')
    .email('Email must be valid and contain `@` symbol'),
  password: Yup.string()
    .required('Please provide your password')
    .min(9, 'Password must be 9 characters or longer ')
    .matches(/[A-Z]/, 'Password should contain at least one capital character')
    .matches(/[a-z]/, 'Password should contain at  least one lowercase char')
    .matches(/[A-Z]/, 'Password should contain at least one uppercase char')
    .matches(/[a-zA-Z]+[^a-zA-Z\s]+/, `
      Password should contain at least 1 number or special char (@,!,#, etc).`)
});

const INITIAL_VALUES = {
  password: '',
  email: '',
  newsletters: '' || true,
  plan: 'premium'
};

export default class SignupForm extends Component {
  buildForm() {
    return ({
      values, errors: { email, password }, touched, isValid, handleReset
    }) => {
      const isEmailHasErrors = touched.email && email;
      const isPasswordHasErrors = touched.password && password;
      return (
        <Form noValidate>
          <div className="form-group row">
            <label htmlFor="email">Enter your email</label>
            <Field
              type="email"
              name="email"
              className={cx('form-control', isEmailHasErrors && 'is-invalid')}
              id="email"
              placeholder="Email"
            />
            {isEmailHasErrors && <InputFeedback error={email} />}
          </div>
          <div className="form-group row">
            <label htmlFor="password"> Enter your password</label>
            <Field
              type="password"
              name="password"
              className={cx('form-control', isPasswordHasErrors &&
                'is-invalid')}
              id="password"
              placeholder="Password"
            />
            {isPasswordHasErrors && <InputFeedback error={password} />}
          </div>
          <div className="form-check row">
            <Field
              type="checkbox"
              name="newsletters"
              className="form-check-input"
              id="newsletters"
              placeholder="Password"
              checked={values.newsletters}
            />
            <label htmlFor="newsletters">
              Would you like to receive our newsletters ?
            </label>
          </div>
          <div className="form-group row">
            <label htmlFor="plan"> Choose your personal plan</label>
            <Field
              component="select"
              name="plan"
              className="form-control"
              id="plan"
            >
              <option value="free">Free</option>
              <option value="premium">Premium</option>
              <option value="gold">Gold</option>
            </Field>
          </div>
          <div className="row">
            <button
              onClick={handleReset}
              className={cx(`${CN}--reset-btn`, 'btn', 'btn-outline-secondary')}
            >
              Reset
            </button>
            <button
              type="submit"
              className="btn btn-primary"
              disabled={!isValid}
            >
              Submit
            </button>
          </div>
        </Form>);
    };
  }
  handleSubmit(values, formikBag) {
    const { resetForm, setErrors } = formikBag;
    let userData = { //eslint-disable-line
      userInfo: {
        email: values.email,
        password: values.password
      },
      plan: values.plan,
      newsletters: values.newsletters
    };
    const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));
    sleep(2000).then(() => {
      if (values.email === 'ivan@gmail.com') {
        const msg = `Email ${values.email} is already taken try another :
         ${values.email}-1 ${values.email}213`;
        setErrors({ email: msg });
      } else {
        console.log(userData); //eslint-disable-line
        resetForm();
      }
    });
  }
  render() {
    const { className } = this.props;
    return (
      <div className={cx(CN, className)}>
        <h2>Sign up</h2>
        <Formik
          initialValues={INITIAL_VALUES}
          onSubmit={this.handleSubmit}
          validationSchema={VALIDATION_SCHEMA}
          render={this.buildForm()}
        />
      </div>);
  }
}

SignupForm.defaultProps = {
  className: ''
};
SignupForm.propTypes = {
  /**
   * className - classes which can be passed from parent
   */
  className: PropTypes.string
};
