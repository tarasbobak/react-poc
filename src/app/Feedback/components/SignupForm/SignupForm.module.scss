.signup-form {
  text-align: left;
  padding: 40px;
  max-width: 400px;

  h2 {
    text-align: center;
  }

  &--reset-btn {
    margin-right: 20px;
  }

}
