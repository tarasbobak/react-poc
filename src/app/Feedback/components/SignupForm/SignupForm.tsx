import React, { Component } from 'react';
import cx from 'classnames';
import { Formik, FormikActions, FormikProps, Form, Field, FieldProps, FormikValues } from 'formik';
import * as Yup from 'yup';
import InputFeedback from './InputFeedback/InputFeedback';
import styles from './SignupForm.module.scss';

interface SignupFormProps {
  className?: string;
}

interface FormValues {
  email: string;
  password: string;
  plan: string;
  newsletters: boolean;
}

const CN = styles['signup-form'];

const VALIDATION_SCHEMA = Yup.object().shape({
  email: Yup.string()
    .required('Please provide your email address, formatted as you@domain.com.')
    .email('Email must be valid and contain `@` symbol'),
  password: Yup.string()
    .required('Please provide your password')
    .min(9, 'Password must be 9 characters or longer ')
    .matches(/[A-Z]/, 'Password should contain at least one capital character')
    .matches(/[a-z]/, 'Password should contain at  least one lowercase char')
    .matches(/[A-Z]/, 'Password should contain at least one uppercase char')
    .matches(
      /[a-zA-Z]+[^a-zA-Z\s]+/,
      `
      Password should contain at least 1 number or special char (@,!,#, etc).`
    )
});

const INITIAL_VALUES = {
  password: '',
  email: '',
  newsletters: '' || true,
  plan: 'premium'
};

export default class SignupForm extends Component<SignupFormProps> {
  protected static defaultProps = {
    className: ''
  };

  public buildForm() {
    return ({ values, errors: { email, password }, touched, isValid, handleReset }: FormikProps<FormValues>) => {
      const isEmailHasErrors = touched.email && email;
      const isPasswordHasErrors = touched.password && password;

      return (
        <Form noValidate={true}>
          <div className="form-group row">
            <label htmlFor="email">Enter your email</label>
            <Field
              type="email"
              name="email"
              className={cx('form-control', isEmailHasErrors && 'is-invalid')}
              id="email"
              placeholder="Email"
            />
            {isEmailHasErrors && <InputFeedback error={email} />}
          </div>
          <div className="form-group row">
            <label htmlFor="password"> Enter your password</label>
            <Field
              type="password"
              name="password"
              className={cx('form-control', isPasswordHasErrors && 'is-invalid')}
              id="password"
              placeholder="Password"
            />
            {isPasswordHasErrors && <InputFeedback error={password} />}
          </div>
          <div className="form-check row">
            <Field
              type="checkbox"
              name="newsletters"
              className="form-check-input"
              id="newsletters"
              placeholder="Password"
              checked={values.newsletters}
            />
            <label htmlFor="newsletters">Would you like to receive our newsletters ?</label>
          </div>
          <div className="form-group row">
            <label htmlFor="plan"> Choose your personal plan</label>
            <Field component="select" name="plan" className="form-control" id="plan">
              <option value="free">Free</option>
              <option value="premium">Premium</option>
              <option value="gold">Gold</option>
            </Field>
          </div>
          <div className="row">
            <button onClick={handleReset} className={cx(`${CN}--reset-btn`, 'btn', 'btn-outline-secondary')}>
              Reset
            </button>
            <button type="submit" className="btn btn-primary" disabled={!isValid}>
              Submit
            </button>
          </div>
        </Form>
      );
    };
  }
  public handleSubmit(values: FormValues, formikActions: FormikActions<FormValues>) {
    const { resetForm, setErrors } = formikActions;
    const userData = {
      userInfo: {
        email: values.email,
        password: values.password
      },
      plan: values.plan,
      newsletters: values.newsletters
    };
    const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

    sleep(2000).then(() => {
      if (values.email === 'ivan@gmail.com') {
        const msg = `Email ${values.email} is already taken try another :
         ${values.email}-1 ${values.email}213`;
        setErrors({ email: msg });
      } else {
        // tslint:disable-next-line
        console.log(userData);
        resetForm();
      }
    });
  }
  public render() {
    const { className } = this.props;

    return (
      <div className={cx(CN, className)}>
        <h2>Sign up</h2>
        <Formik
          initialValues={INITIAL_VALUES}
          onSubmit={this.handleSubmit}
          validationSchema={VALIDATION_SCHEMA}
          render={this.buildForm()}
        />
      </div>
    );
  }
}
