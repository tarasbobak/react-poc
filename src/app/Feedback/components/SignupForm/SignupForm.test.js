import React from 'react';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import SignupForm from './SignupForm';

configure({ adapter: new Adapter() });

describe('Test header component to render properly with properties', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      handleSubmit: () => {},
      buildForm: () => {}
    };
    wrapper = shallow(<SignupForm {...props} />);
  });

  it('should be rendered properly with different properties', () => {
    const tree = renderer.create(
      <SignupForm onSubmit={() => {}} render={() => {}} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be rendered properly', () => {
    const tree = renderer.create(
      <SignupForm>
        test
      </SignupForm>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('first child form must be `<h2/>`', () => {
    expect(
      wrapper.childAt(0).type()
    ).toBe('h2');
  });

  it('onSubmit method should be defined', () => {
    expect(
      wrapper.childAt(1).props().onSubmit
    ).toBeDefined();
  });

  it('render method should be defined', () => {
    expect(
      wrapper.childAt(1).props().render
    ).toBeDefined();
  });

  it('onSubmit method should be defined', () => {
    expect(
      wrapper.childAt(1).props().onSubmit
    ).toBeDefined();
  });

  it('SignupForm component must have 2 children', () => {
    expect(
      wrapper.children().length
    ).toBe(2);
  });

});

