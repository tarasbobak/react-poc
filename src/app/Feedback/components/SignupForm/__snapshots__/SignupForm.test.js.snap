// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test header component to render properly with properties should be rendered properly 1`] = `
<div
  className="signup-form"
>
  <h2>
    Sign up
  </h2>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="form-group row"
    >
      <label
        htmlFor="email"
      >
        Enter your email
      </label>
      <input
        className="form-control"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <div
      className="form-group row"
    >
      <label
        htmlFor="password"
      >
         Enter your password
      </label>
      <input
        className="form-control"
        id="password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <div
      className="form-check row"
    >
      <input
        checked={true}
        className="form-check-input"
        id="newsletters"
        name="newsletters"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Password"
        type="checkbox"
        value={undefined}
      />
      <label
        htmlFor="newsletters"
      >
        Would you like to receive our newsletters ?
      </label>
    </div>
    <div
      className="form-group row"
    >
      <label
        htmlFor="plan"
      >
         Choose your personal plan
      </label>
      <select
        className="form-control"
        id="plan"
        name="plan"
        onBlur={[Function]}
        onChange={[Function]}
        value="premium"
      >
        <option
          value="free"
        >
          Free
        </option>
        <option
          value="premium"
        >
          Premium
        </option>
        <option
          value="gold"
        >
          Gold
        </option>
      </select>
    </div>
    <div
      className="row"
    >
      <button
        className="signup-form--reset-btn btn btn-outline-secondary"
        onClick={[Function]}
      >
        Reset
      </button>
      <button
        className="btn btn-primary"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`Test header component to render properly with properties should be rendered properly with different properties 1`] = `
<div
  className="signup-form"
>
  <h2>
    Sign up
  </h2>
  <form
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="form-group row"
    >
      <label
        htmlFor="email"
      >
        Enter your email
      </label>
      <input
        className="form-control"
        id="email"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <div
      className="form-group row"
    >
      <label
        htmlFor="password"
      >
         Enter your password
      </label>
      <input
        className="form-control"
        id="password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </div>
    <div
      className="form-check row"
    >
      <input
        checked={true}
        className="form-check-input"
        id="newsletters"
        name="newsletters"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Password"
        type="checkbox"
        value={undefined}
      />
      <label
        htmlFor="newsletters"
      >
        Would you like to receive our newsletters ?
      </label>
    </div>
    <div
      className="form-group row"
    >
      <label
        htmlFor="plan"
      >
         Choose your personal plan
      </label>
      <select
        className="form-control"
        id="plan"
        name="plan"
        onBlur={[Function]}
        onChange={[Function]}
        value="premium"
      >
        <option
          value="free"
        >
          Free
        </option>
        <option
          value="premium"
        >
          Premium
        </option>
        <option
          value="gold"
        >
          Gold
        </option>
      </select>
    </div>
    <div
      className="row"
    >
      <button
        className="signup-form--reset-btn btn btn-outline-secondary"
        onClick={[Function]}
      >
        Reset
      </button>
      <button
        className="btn btn-primary"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
